### DISCUSSION TITLE
Feature Request: GPU support – CUDA / Metal

### DISCUSSION THREAD

**Discussion by prmelehan** [duckdb/duckdb#8675] [Ideas]
In the past month, I have extensively used DuckDB in various projects, finding it to be an outstanding software solution. My work with DuckDB has expanded to include local usage with large datasets (25-50GB). However, I've encountered difficulties in handling workloads that require joining or grouping large amounts of distinct values.

I believe that the inclusion of native CUDA and Metal support would significantly enhance DuckDB's capabilities. Metal's integration with Apple Silicon chips, which minimizes memory transfer overhead between the CPU and GPU, would be particularly beneficial. Similarly, CUDA support would enable faster processing of larger datasets in platforms like AWS or GCP.

If these features were included by default, or at least as installable extensions, it would elevate DuckDB's performance to new levels, making it capable of handling tasks at what could be described as light speed. :)

**Comment by l1t1** [comment]
what are the fields of your db that can benifit from gpu?

**Comment by szarnyasg** [comment]
GPUs have limited support for variable-sized data, so even handling strings is challenging not to mention lists. The transfer times to and from the GPU also limit potential performance gains. Larger GPUs are also very expensive, especially when used in the cloud. Consequently, very few databases use GPUs: https://dbdb.io/browse?q=gpu

We are also not planning to support GPUs anytime soon.

Regarding your performance problem, joins and group-bys parallelize nicely on many-core CPUs. Did you try memory-optimized cloud instances with ~5-10GB memory/CPU and NVMe disks, e.g. an `r6id.4xlarge` (Intel Xeon cores) or `r6gd.4xlarge` (Graviton cores)?

**Reply by prmelehan** [reply]
Out of curiosity, what are the limitations around strings and lists? Is there a resource where I could read more?

**Comment by leventov** [comment]
@prmelehan for Metal support, I believe you should preach to Apple engineers to make an Apache Arrow library (a-la https://github.com/ml-explore/mlx), and leverage DuckDB's integration with Arrow Streaming. Some supporting groundwork was done in Arrow here: https://github.com/apache/arrow/pull/34972

**Reply by prmelehan** [reply]
Interesting. So just adding Arrow support would allow for a massive speedup? No changes to the query execution engine required?

https://arrow.apache.org/docs/format/CDeviceDataInterface.html

I found this on CDeviceDataInterface. This is not my expertise, but it does mention a ARROW_DEVICE_METAL flag? Could this be of use?

**Comment by alephonea** [comment]
Reopening this thread: has DuckDB reconsidered GPU support? Would be a nice feature for data analytics to use GPU for acceleration.

There has been some progress on features - cuDF now supports string handling:
https://docs.rapids.ai/api/cudf/nightly/user_guide/api_docs/string_handling/

**Reply by prmelehan** [reply]
I'd love to see some support here. Happy to implement stuff myself, but would need a bit of pointing in the right direction.

GPU support mostly makes sense for Mac at this point. NVIDIA cards are expensive and not many have them inside of their PCs.

All Macs have Metal support (with up to 80% of RAM as VRAM), and for local data analytics would be amazing.

**Reply by alephonea** [reply]
Just followed up on your website to discuss.

**Comment by singhpratech** [comment]
The route leventov described above now exists as a library, so a data point for this thread, with
the numbers that support szarnyasg's point as well as the ones that do not. ArrowMetal
(https://github.com/singhpratech/ArrowMetal, Apache-2.0) runs Apache Arrow compute on the Apple GPU
through Metal. DuckDB needs no change: it hands a result out over the Arrow C Data Interface, and on
Apple silicon the CPU and GPU share one memory, so a Metal buffer can wrap an Arrow buffer in place
(a test asserts the import is zero-copy). What does cost is DuckDB's own chunking: at 50M rows one
BIGINT column arrives as 407 chunks, and making it one contiguous buffer is a real copy, 26.6 ms of a
52.1 ms crossing.

```python
import duckdb, arrowmetal as am
con = duckdb.connect()
cols = am.from_duckdb(con.sql("select k, v from read_parquet('trades/*.parquet')"))  # {name: MetalArray}
gb = am.group_by([cols["k"]])
rel = am.to_duckdb(con, "gpu_totals", {"k": gb.keys()[0], "total": gb.sum(cols["v"])})  # back as a view
```

Measured on an M4 Max, DuckDB 1.5.5, 50M rows, same in-memory table on both sides, answers checked
equal before timing (docs/DUCKDB.md has the full table and the script). Ratio is DuckDB's wall time
over the GPU's; "one-shot" includes the crossing, "resident" is the column already on the GPU:

| operation | DuckDB ms (CPU-ms) | GPU resident ms | one-shot | resident |
|---|---:|---:|---:|---:|
| filter + sum, fused | 1.1 (13.6) | 1.1 | 0.0x | 1.0x |
| group-by sum, 1k keys | 7.6 (112) | 2.1 | 0.1x | 3.6x |
| group-by sum, 100k keys | 71.4 (1,056) | 4.5 | 1.1x | 16.0x |
| top_k(100) | 5.7 (12) | 9.7 | 0.1x | 0.6x |
| sort | 39.2 (570) | 17.6 | 0.7x | 2.2x |
| count(s like '%user1%') | 47.4 (719) | 5.1 | 0.6x | 9.3x |
| avg(f) | 2.8 (41) | 2.5 | 0.1x | 1.1x |

So for a single query the transfer cost decides it, as szarnyasg said: DuckDB is ahead on every row
but the 100k-key group-by. The case for the GPU is a column that stays resident across several
operations, where the heavy rows (high-cardinality group-by, string scan, sort) are 2x to 16x, and
the CPU-ms column: DuckDB's 100k-key group-by uses about fifteen cores for 71 ms, the GPU's uses
0.6 CPU-ms. Strings do work (the `like` row is utf8 with nulls; the Arrow layout carries the
offsets). There is also a loadable extension for SQL-only use; in 0.1.0 it is behind DuckDB's own
SQL, and the doc says why.
